source-postgres/capture.go

```diff
@@ -0,0 +1,570 @@
+		// to the zero-length range after the previous range.
+		//
+		// Whether or not this is the case, `lastRange` will end up referring to some range
+		// with the same `ScannedLSN` as our current transaction, which will be extended by
+		// reading further chunks from the table.
+		lastRange := &tableState.ScanRanges[len(tableState.ScanRanges)-1]
+		if lastRange.ScannedLSN != snapshot.TransactionLSN() {
+			tableState.ScanRanges = append(tableState.ScanRanges, TableRange{
+				ScannedLSN: snapshot.TransactionLSN(),
+				EndKey:     lastRange.EndKey,
+			})
+			lastRange = &tableState.ScanRanges[len(tableState.ScanRanges)-1]
+		}
+
+		// This is the core operation where we spend most of our time when scanning
+		// table contents: reading the next chunk from the table and extending the
+		// endpoint of `lastRange`.
+		count, nextKey, err := table.ScanFrom(ctx, lastRange.EndKey, c.HandleChangeEvent)
+		if err != nil {
+			return errors.Wrap(err, "error processing snapshot events")
+		}
+		if count == 0 {
+			tableState.Mode = ScanStateActive
+		} else {
+			lastRange.EndKey = nextKey
+		}
+		if err := c.EmitState(c.State); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// This is the "payoff" of all the setup work we did in `UpdateState` and `ScanTables`,
+// in that once we reach this state we'll sit here indefinitely, awaiting change events
+// from PostgreSQL via logical replication and relaying these changes downstream ASAP.
+//
+// Each time we receive a replication event from PostgreSQL it will be decoded and then
+// passed off to `c.HandleReplicationEvent` for further processing.
+func (c *Capture) StreamChanges(ctx context.Context) error {
+	stream, err := StartReplication(ctx, c.connRepl, c.Config.SlotName, c.Config.PublicationName, c.State.CurrentLSN)
+	if err != nil {
+		return errors.Wrap(err, "unable to start replication stream")
+	}
+	defer stream.Close(ctx)
+	if err := stream.Process(ctx, c.HandleReplicationEvent); err != nil {
+		return errors.Wrap(err, "error processing replication events")
+	}
+	logrus.Warnf("replication streaming terminated without error")
+	return nil
+}
+
+func (c *Capture) HandleReplicationEvent(evt *ChangeEvent) error {
+	// When in non-tailing mode, reset the shutdown watchdog whenever an event happens.
+	if c.watchdog != nil {
+		wdtDuration := time.Duration(c.Config.PollTimeoutSeconds * float64(time.Second))
+		c.watchdog.Reset(wdtDuration)
+	}
+
+	// TODO(wgd): There will be a "Commit" event after every transaction, so if the DB
+	// receives a bunch of single-row INSERTs we'll emit one state update per row added.
+	// Is this too much?
+	//
+	// TODO(wgd): At some point during replication event processing we should be far
+	// enough past the initial table scan that ideally we want to delete all the scan
+	// ranges and drop into the fast-path of not doing any range comparison. Is there
+	// ever a point where we can definitively say "Yes, we're fully caught up now"?
```

Comment:
If we sample an LSN M _after_ ScanTables completes, read through M, and later see an LSN N < M, aren't we still OK?

The reason why I would think so is that, while N itself might be < M, N's _commit_ must be > M. So there's still no need to compare N to the ranges because it could not have been observed by ScanTables.

> I'm not sure if that's preferable to the current watchdog-timeout approach.

The advantage would simply be that it's faster to detect the shutdown condition and exit.
Workflows like `flowctl develop --poll` that are tracing the exit and, from there, propagations of data through the graph will run faster end-to-end.